File: frontend/common/modules/providers/lib/mutationqueue.js
import { debounce, uniqueStore } from '@/modules/util.js'
import { cache, caches } from '@/modules/cache.js'
import { profiles } from '@/modules/settings.js'
import { writable } from 'simple-store-svelte'
import Debug from 'debug'
const debug = Debug('ui:mutationqueue')

/**
 * @typedef {'entry' | 'delete' | 'favourite'} MutationType
 */

/**
 * @typedef {object} QueuedMutation
 * @property {MutationType} type
 * @property {number} mediaId
 * @property {Record<string, any>} variables
 * @property {any} result
 * @property {number | null} progressBefore Progress value at the time of queueing, used for stale-write detection.
 * @property {boolean} executed True if the API call already succeeded this session; false if offline/pending.
 * @property {number} queuedAt Unix timestamp (ms) when this mutation was enqueued.
 */

export class MutationQueue {
  /** @type {'AniList' | 'MyAnimeList'} */
  #provider
  /** @type {import('simple-store-svelte').Writable<QueuedMutation[]>} */
  #queue = writable([])
  /** @type {boolean} True while the user lists are being fetched, used to decide whether to queue or apply immediately. */
  isFetchingList = false
  /**
   * @type {{ AniList: { perMinute: number, delayMs: number }, MyAnimeList: { perMinute: number, delayMs: number } }}
   * Rate limits per provider.
   */
  RATE_LIMITS = {
    AniList:     { perMinute: 8,  delayMs: 500 },
    MyAnimeList: { perMinute: 30, delayMs: 1_500 }
  }

  /**
   * @param {string} cacheKey Cache key used to persist the queue across sessions.
   * @param {string | null} [malAuth] If user is logged into MAL, tightens the MyAnimeList rate limit to 15 req/min.
   */
  constructor(cacheKey, malAuth) {
    this.#provider = cacheKey === 'syncQueueAni' ? 'AniList' : 'MyAnimeList'
    if (malAuth) this.RATE_LIMITS.MyAnimeList = { perMinute: 15, delayMs: 3_000 }
    try {
      const stored = cache.getEntry(caches.GENERAL, cacheKey)
      if (Array.isArray(stored) && stored.length) {
        this.#queue.value = stored
        debug(`[${this.#provider}] loaded ${this.#queue.value.length} persisted mutation(s)`)
      }
    } catch (error) {
      debug(`[${this.#provider}] failed to load:`, error)
      this.#queue.value = []
    }

    const updateCache = debounce((value) => {
      debug(`[${this.#provider}] the number of mutations in the queue have changed, saving to cache...`)
      cache.setEntry(caches.GENERAL, cacheKey, value.filter(mutation => !mutation.executed))
    }, 20)
    uniqueStore(this.#queue).subscribe(value => updateCache(value))
  }

  /**
   * Adds a mutation to the queue.
   * Duplicate favourite toggles for the same media cancel each other out, while duplicate entry/delete mutations use last-write-wins, preserving the original progressBefore and queuedAt.
   * Returns false if executed=true and the list is not currently being fetched, signaling the caller to apply the result immediately instead.
   *
   * @param {MutationType} type
   * @param {number} mediaId
   * @param {Record<string, any>} variables
   * @param {any} result
   * @param {number | null} [progressBefore] Progress before this mutation, used for stale-write detection.
   * @param {boolean} [executed] Whether the API call has already been made. Defaults to true.
   * @returns {boolean} false if the caller should apply the result immediately, true if queued.
   */
  enqueue(type, mediaId, variables, result, progressBefore = null, executed = true) {
    if (executed && !this.isFetchingList) return false
    let updatedVariables = { ...(variables ? variables : {}) }
    const userToken = updatedVariables.token
    if (userToken) {
      const profile = profiles.value.find(profile => profile?.token === userToken && profile?.viewer?.data?.Viewer?.id)
      if (profile) {
        delete updatedVariables.refresh_in
        delete updatedVariables.token
        updatedVariables.tokenUserId = profile.viewer.data.Viewer.id
      }
    }
    const mutation = { type, mediaId, variables: updatedVariables, result, progressBefore, executed, queuedAt: Date.now() }
    const queue = [...this.#queue.value]

    if (type === 'favourite') {
      const index = queue.findIndex(mutation => mutation.type === 'favourite' && mutation.mediaId === mediaId)
      if (index !== -1) {
        // Two toggles cancel out
        debug(`[${this.#provider}] favourite toggles for media ${mediaId} cancel out`)
        queue.splice(index, 1)
        this.#queue.value = queue
        return true
      }
    } else {
      const index = queue.findIndex(mutation => mutation.type === type && mutation.mediaId === mediaId)
      if (index !== -1) {
        // Last write wins, but preserve original progressBefore and queuedAt
        debug(`[${this.#provider}] replacing ${type} for media ${mediaId}`)
        queue[index] = { ...mutation, progressBefore: queue[index].progressBefore, queuedAt: queue[index].queuedAt }
        this.#queue.value = queue
        return true
      }
    }

    debug(`[${this.#provider}] enqueuing ${type} for media ${mediaId} (executed=${executed})`)
    this.#queue.value = [...queue, mutation]
    return true
  }

  /**
   * Returns the pre-mutation progress for a queued entry mutation, or null if none exists.
   * Used to avoid reading an already-optimistically-mutated cache value as the baseline.
   *
   * @param {number} mediaId
   * @returns {number | null}
   */
  getProgressBefore(mediaId) {
    return this.#queue.value.find(mutation => mutation.type === 'entry' && mutation.mediaId === mediaId)?.progressBefore ?? null
  }

  /** @returns {boolean} Whether there are any pending mutations in the queue. */
  get hasPending() {
    return this.#queue.value.length > 0
  }

  /**
   * Flushes the queue after userLists.value has been set.
   * All executed mutations are applied to the local cache immediately via applyFn, while offline mutations are sent to the API via executeFn with rate limiting.
   * Offline mutations remain in the persisted queue until they succeed, so an app restart mid-flush will retry them on next launch.
   *
   * @param {import('../anilist/al.d.ts').MediaListCollection | import('../myanimelist/mal.d.ts').MediaList | null} resolvedLists Fresh list data from the API, used for stale-write validation.
   * @param {(mutation: QueuedMutation) => Promise<void>} applyFn Applies a mutation to the local cache.
   * @param {(mutation: QueuedMutation) => Promise<void>} executeFn Sends an offline mutation to the API.
   */
  async flush(resolvedLists, applyFn, executeFn) {
    if (!this.#queue.value.length) return
    const snapshot = [...this.#queue.value]
    this.#queue.value = snapshot.filter(mutation => !mutation.executed) // Clear only the executed (session-only race-condition) mutations upfront, offline mutations stay in the queue until they successfully execute.
    debug(`[${this.#provider}] flushing ${snapshot.length} mutation(s)`)
    const offline = snapshot.filter(mutation => !mutation.executed)
    // Apply all validated mutations to local cache immediately so the UI is up to date regardless of how long the rate-limited API calls take.
    for (const mutation of [...snapshot.filter(mutation => mutation.executed).reverse(), ...offline]) {
      if (!this.#validate(mutation, resolvedLists)) continue
      if (mutation.executed) await applyFn(this.#resolveTokenFields(mutation))
      else if (mutation.result != null) await applyFn(this.#resolveTokenFields(mutation))
    }
    await this.#executeRateLimited(offline, resolvedLists, executeFn)
  }

  /**
   * Sends offline mutations to the API one-by-one, respecting the providers rate limit.
   * Each mutation is removed from the persisted queue only after it succeeds.
   * Failed mutations are left in the queue to be retried on the next flush.
   *
   * @param {QueuedMutation[]} mutations
   * @param {import('../anilist/al.d.ts').MediaListCollection | import('../myanimelist/mal.d.ts').MediaList | null} resolvedLists
   * @param {(mutation: QueuedMutation) => Promise<void>} executeFn
   */
  async #executeRateLimited(mutations, resolvedLists, executeFn) {
    const { perMinute, delayMs } = this.RATE_LIMITS[this.#provider]
    let executedThisWindow = 0
    let windowStart = Date.now()
    for (const mutation of mutations) {
      if (!this.#validate(mutation, resolvedLists)) {
        // Not valid... Discard and remove from the persisted queue
        this.#queue.value = this.#queue.value.filter(_mutation => !(_mutation.type === mutation.type && _mutation.mediaId === mutation.mediaId && _mutation.queuedAt === mutation.queuedAt))
        continue
      }

      // If we have hit the rate limit for this window, wait until the window resets
      if (executedThisWindow >= perMinute) {
        const elapsed = Date.now() - windowStart
        const waitMs = 60_000 - elapsed
        if (waitMs > 0) {
          debug(`[${this.#provider}] rate limit reached (${perMinute}/min), waiting ${waitMs}ms`)
          await new Promise(resolve => setTimeout(resolve, waitMs).unref?.())
        }
        executedThisWindow = 0
        windowStart = Date.now()
      }

      try {
        debug(`[${this.#provider}] executing offline ${mutation.type} for media ${mutation.mediaId}`)
        await executeFn(this.#resolveTokenFields(mutation))
        // Successfully executed, remove from the persisted queue
        this.#queue.value = this.#queue.value.filter(_mutation => !(_mutation.type === mutation.type && _mutation.mediaId === mutation.mediaId && _mutation.queuedAt === mutation.queuedAt))
        executedThisWindow++
      } catch (error) {
        debug(`[${this.#provider}] offline ${mutation.type} for media ${mutation.mediaId} failed, keeping in queue:`, error)
        // Leave it in the queue, it will be retried on next flush
      }
      // Small delay between requests to avoid hammering the API
      if (delayMs > 0) await new Promise(resolve => setTimeout(resolve, delayMs).unref?.())
    }
  }

  /**
   * Validates a mutation against fresh server data to detect stale writes, delete and favourite mutations are always considered valid.
   * An entry mutation is invalid if the servers progress has already moved past the baseline recorded at queue time, indicating a newer write won.
   *
   * @param {QueuedMutation} mutation
   * @param {import('../anilist/al.d.ts').MediaListCollection | import('../myanimelist/mal.d.ts').MediaList | null} resolvedLists
   * @returns {boolean}
   */
  #validate(mutation, resolvedLists) {
    if (mutation.type === 'delete' || mutation.type === 'favourite') return true
    if (mutation.type === 'entry') {
      const targetProgress = mutation.executed ? mutation.result?.progress : mutation.variables?.episode
      if (targetProgress == null) return true
      const freshProgress = this.#findProgress(resolvedLists, mutation.mediaId)
      if (freshProgress == null) return true
      const progressBefore = mutation.progressBefore ?? freshProgress
      if (freshProgress > progressBefore) {
        debug(`[${this.#provider}] discard ${mutation.mediaId}: server (${freshProgress}) moved past start (${progressBefore})`)
        return false
      }
      return true
    }
    return true
  }

  /**
   * Looks up the current watch progress for a media entry from either an AniList or MAL list response.
   *
   * @param {import('../anilist/al.d.ts').MediaListCollection | import('../myanimelist/mal.d.ts').MediaList | null} lists
   * @param {number} mediaId
   * @returns {number | null}
   */
  #findProgress(lists, mediaId) {
    if (!lists) return null
    // AniList shape: { lists: [{ entries: [{ media: { id, mediaListEntry: { progress } } }] }] }
    if (lists.lists) {
      for (const list of lists.lists) {
        const entry = list.entries?.find(e => e.media?.id === mediaId)
        if (entry) return entry.media?.mediaListEntry?.progress ?? null
      }
      return null
    }
    // MAL shape: [{ node: { id, my_list_status: { num_episodes_watched } } }]
    if (Array.isArray(lists)) {
      const entry = lists.find(e => e.node?.id === mediaId)
      return entry?.node?.my_list_status?.num_episodes_watched ?? null
    }
    return null
  }

  /**
   * Restores the token and refresh_in fields to a mutations variables before execution.
   * At enqueue time these are replaced with tokenUserId to avoid persisting sensitive token data.
   *
   * @param {QueuedMutation} mutation
   * @returns {QueuedMutation}
   */
  #resolveTokenFields(mutation) {
    const { tokenUserId } = mutation.variables ?? {}
    if (!tokenUserId) return mutation
    const profile = profiles.value.find(profile => profile?.viewer?.data?.Viewer?.id === tokenUserId)
    if (!profile) return mutation
    const variables = { ...mutation.variables, token: profile.token, refresh_in: profile.refresh_in }
    delete variables.tokenUserId
    return { ...mutation, variables }
  }
}

File: app/lib/infrastructure/tracking/mutation_queue.dart
// ignore_for_file: prefer_initializing_formals

/// Offline-safe mutation queue, ported from the redo branch's
/// `providers/lib/mutationqueue.js`.
///
/// Semantics preserved:
/// - persisted across sessions (only NON-executed mutations are persisted);
/// - types entry | delete | favourite;
/// - rate limits: AniList 8/min with 500 ms spacing; MAL 30/min with 1500 ms
///   (tightened to 15/min with 3000 ms when MAL is the authenticated
///   provider);
/// - duplicate favourite toggles for the same media cancel each other out;
/// - duplicate entry/delete are last-write-wins, preserving the original
///   `progressBefore` and `queuedAt`;
/// - `progressBefore` powers stale-write detection at flush time: an entry
///   mutation is discarded when the server's progress already moved past the
///   baseline recorded at queue time;
/// - tokens are NEVER persisted inside variables — they are replaced with a
///   `tokenUserId` reference and restored just before execution;
/// - the queue is flushed on the reconnect signal and after list reads
///   settle (the caller invokes [flush] at those moments; `isFetchingList`
///   guards the race).
library;

import '../../domain/ports/query_cache.dart';
import '../network/transport.dart';
import 'sync_rules.dart' show TrackingProvider;

enum MutationType { entry, delete, favourite }

class QueuedMutation {
  const QueuedMutation({
    required this.type,
    required this.mediaId,
    required this.variables,
    this.result,
    this.progressBefore,
    required this.executed,
    required this.queuedAt,
  });

  factory QueuedMutation.fromJson(Map<String, dynamic> json) => QueuedMutation(
    type: MutationType.values.byName(json['type'] as String),
    mediaId: (json['mediaId'] as num).toInt(),
    variables: (json['variables'] as Map?)?.cast<String, dynamic>() ?? const {},
    result: (json['result'] as Map?)?.cast<String, dynamic>(),
    progressBefore: (json['progressBefore'] as num?)?.toInt(),
    executed: json['executed'] as bool? ?? false,
    queuedAt: DateTime.fromMillisecondsSinceEpoch(
      (json['queuedAt'] as num).toInt(),
    ),
  );

  final MutationType type;
  final int mediaId;

  /// Mutation variables. Never carries `token`/`refresh_in` while queued —
  /// only a `tokenUserId` reference.
  final Map<String, dynamic> variables;

  /// The API result when [executed] (used to re-apply to the local cache).
  final Map<String, dynamic>? result;

  /// Progress at queue time — the stale-write baseline.
  final int? progressBefore;

  /// True when the API call already succeeded this session; false when the
  /// mutation is waiting to be sent (offline).
  final bool executed;

  final DateTime queuedAt;

  QueuedMutation copyWith({
    Map<String, dynamic>? variables,
    int? progressBefore,
    DateTime? queuedAt,
  }) => QueuedMutation(
    type: type,
    mediaId: mediaId,
    variables: variables ?? this.variables,
    result: result,
    progressBefore: progressBefore ?? this.progressBefore,
    executed: executed,
    queuedAt: queuedAt ?? this.queuedAt,
  );

  Map<String, dynamic> toJson() => {
    'type': type.name,
    'mediaId': mediaId,
    'variables': variables,
    if (result != null) 'result': result,
    if (progressBefore != null) 'progressBefore': progressBefore,
    'executed': executed,
    'queuedAt': queuedAt.millisecondsSinceEpoch,
  };

  bool _same(QueuedMutation other) =>
      other.type == type &&
      other.mediaId == mediaId &&
      other.queuedAt == queuedAt;
}

/// Restored credentials for a queued mutation ([MutationQueue.credentialsOf]).
class MutationCredentials {
  const MutationCredentials(this.token, {this.refreshIn});

  final String token;

  /// Epoch seconds of the MAL refresh deadline (`refresh_in`).
  final int? refreshIn;
}

class MutationRateLimit {
  const MutationRateLimit({required this.perMinute, required this.spacing});

  final int perMinute;
  final Duration spacing;
}

typedef PersistQueue = Future<void> Function(List<QueuedMutation> pending);

class MutationQueue {
  /// [persist] receives the non-executed mutations on every change; pass a
  /// callback of your own, or use [MutationQueue.fromCache] for the old
  /// `GENERAL:syncQueueAni` / `syncQueueMal` persistence.
  ///
  /// [malAuthenticated] tightens the MyAnimeList limit to 15/min, 3000 ms —
  /// the old app did this when MAL was the signed-in provider (its own
  /// account traffic already consumes MAL quota).
  MutationQueue({
    required this.provider,
    PersistQueue? persist,
    List<QueuedMutation> initial = const [],
    bool malAuthenticated = false,
    Clock clock = const SystemClock(),
    int? Function(String token)? tokenUserIdOf,
    MutationCredentials? Function(int userId)? credentialsOf,
  }) : _persist = persist,
       _clock = clock,
       _tokenUserIdOf = tokenUserIdOf,
       _credentialsOf = credentialsOf,
       _queue = [...initial],
       rateLimit = provider == TrackingProvider.aniList
           ? const MutationRateLimit(
               perMinute: 8,
               spacing: Duration(milliseconds: 500),
             )
           : malAuthenticated
           ? const MutationRateLimit(
               perMinute: 15,
               spacing: Duration(milliseconds: 3000),
             )
           : const MutationRateLimit(
               perMinute: 30,
               spacing: Duration(milliseconds: 1500),
             );

  /// Loads the persisted queue from the old cache location
  /// (`general:syncQueueAni` / `general:syncQueueMal`) and keeps it in sync.
  static Future<MutationQueue> fromCache(
    QueryCache cache, {
    required TrackingProvider provider,
    bool malAuthenticated = false,
    Clock clock = const SystemClock(),
    int? Function(String token)? tokenUserIdOf,
    MutationCredentials? Function(int userId)? credentialsOf,
  }) async {
    final key = provider == TrackingProvider.aniList
        ? 'syncQueueAni'
        : 'syncQueueMal';
    var initial = const <QueuedMutation>[];
    try {
      final stored = await cache.read(
        CacheStores.general,
        key,
        ignoreExpiry: true,
      );
      final list = stored?.value['mutations'] as List?;
      if (list != null) {
        initial = [
          for (final item in list.whereType<Map>())
            QueuedMutation.fromJson(item.cast<String, dynamic>()),
        ];
      }
    } on Object {
      initial = const [];
    }
    return MutationQueue(
      provider: provider,
      initial: initial,
      malAuthenticated: malAuthenticated,
      clock: clock,
      tokenUserIdOf: tokenUserIdOf,
      credentialsOf: credentialsOf,
      persist: (pending) => cache.write(CacheStores.general, key, {
        'mutations': [for (final m in pending) m.toJson()],
      }),
    );
  }

  final TrackingProvider provider;
  final MutationRateLimit rateLimit;
  final PersistQueue? _persist;
  final Clock _clock;
  final int? Function(String token)? _tokenUserIdOf;
  final MutationCredentials? Function(int userId)? _credentialsOf;

  final List<QueuedMutation> _queue;
  Future<void> _persistFuture = Future.value();

  /// True while user lists are being fetched — decides whether an executed
  /// mutation must be queued (to be re-applied after the fetch lands) or can
  /// be applied immediately.
  bool isFetchingList = false;

  bool get hasPending => _queue.isNotEmpty;

  List<QueuedMutation> get pending => List.unmodifiable(_queue);

  /// Completes when the latest persistence write has settled (test hook).
  Future<void> get persisted => _persistFuture;

  void _schedulePersist() {
    final persist = _persist;
    if (persist == null) return;
    final pending = [
      for (final m in _queue)
        if (!m.executed) m,
    ];
    _persistFuture = _persistFuture.then((_) => persist(pending));
  }

  /// Strips credentials from the variables, replacing them with a
  /// `tokenUserId` reference so tokens are never persisted.
  Map<String, dynamic> _stripToken(Map<String, dynamic> variables) {
    final token = variables['token'];
    if (token is! String) return variables;
    final userId = _tokenUserIdOf?.call(token);
    final safe = {...variables}
      ..remove('token')
      ..remove('refresh_in');
    if (userId != null) safe['tokenUserId'] = userId;
    return safe;
  }

  /// Restores token/refresh_in from the `tokenUserId` reference right before
  /// execution.
  QueuedMutation resolveTokenFields(QueuedMutation mutation) {
    final userId = mutation.variables['tokenUserId'];
    if (userId is! num) return mutation;
    final credentials = _credentialsOf?.call(userId.toInt());
    if (credentials == null) return mutation;
    final variables = {...mutation.variables}..remove('tokenUserId');
    variables['token'] = credentials.token;
    if (credentials.refreshIn != null) {
      variables['refresh_in'] = credentials.refreshIn;
    }
    return mutation.copyWith(variables: variables);
  }

  /// Adds a mutation. Returns false when [executed] is true and no list
  /// fetch is in flight — the caller should apply the result immediately
  /// instead of queueing.
  bool enqueue(
    MutationType type,
    int mediaId,
    Map<String, dynamic> variables, {
    Map<String, dynamic>? result,
    int? progressBefore,
    bool executed = true,
  }) {
    if (executed && !isFetchingList) return false;
    final mutation = QueuedMutation(
      type: type,
      mediaId: mediaId,
      variables: _stripToken(variables),
      result: result,
      progressBefore: progressBefore,
      executed: executed,
      queuedAt: _clock.now(),
    );

    if (type == MutationType.favourite) {
      final index = _queue.indexWhere(
        (m) => m.type == MutationType.favourite && m.mediaId == mediaId,
      );
      if (index != -1) {
        // Two toggles cancel out.
        _queue.removeAt(index);
        _schedulePersist();
        return true;
      }
    } else {
      final index = _queue.indexWhere(
        (m) => m.type == type && m.mediaId == mediaId,
      );
      if (index != -1) {
        // Last write wins, preserving the original baseline and queue time.
        _queue[index] = mutation.copyWith(
          progressBefore: _queue[index].progressBefore,
          queuedAt: _queue[index].queuedAt,
        );
        _schedulePersist();
        return true;
      }
    }

    _queue.add(mutation);
    _schedulePersist();
    return true;
  }

  /// The pre-mutation progress for a queued entry mutation — used so a
  /// second write while offline keeps the original baseline instead of
  /// reading the already-optimistically-patched cache.
  int? progressBeforeOf(int mediaId) {
    for (final m in _queue) {
      if (m.type == MutationType.entry && m.mediaId == mediaId) {
        return m.progressBefore;
      }
    }
    return null;
  }

  /// Flushes after a list read settles or on reconnect.
  ///
  /// Executed mutations are re-applied to the local cache via [apply] (they
  /// raced the list fetch); offline mutations are also applied when they
  /// carry a result, then sent to the API via [execute] under the provider
  /// rate limit. A mutation leaves the persisted queue only after its API
  /// call succeeds; failures stay queued for the next flush.
  ///
  /// [freshProgressOf] reports the server's current progress for a media
  /// (from the just-fetched lists) and powers stale-write detection.
  Future<void> flush({
    int? Function(int mediaId)? freshProgressOf,
    required Future<void> Function(QueuedMutation mutation) apply,
    required Future<void> Function(QueuedMutation mutation) execute,
  }) async {
    if (_queue.isEmpty) return;
    final snapshot = [..._queue];
    // Executed (session-only race) mutations clear upfront; offline ones
    // stay persisted until they actually go through.
    _queue.removeWhere((m) => m.executed);
    _schedulePersist();

    final offline = [
      for (final m in snapshot)
        if (!m.executed) m,
    ];
    final executedNewestFirst = [
      for (final m in snapshot.reversed)
        if (m.executed) m,
    ];

    // Re-apply to the local cache immediately so the UI is current no
    // matter how long the rate-limited API calls take.
    for (final mutation in [...executedNewestFirst, ...offline]) {
      if (!_validate(mutation, freshProgressOf)) continue;
      if (mutation.executed || mutation.result != null) {
        await apply(resolveTokenFields(mutation));
      }
    }

    await _executeRateLimited(offline, freshProgressOf, execute);
  }

  Future<void> _executeRateLimited(
    List<QueuedMutation> mutations,
    int? Function(int mediaId)? freshProgressOf,
    Future<void> Function(QueuedMutation mutation) execute,
  ) async {
    var executedThisWindow = 0;
    var windowStart = _clock.now();
    for (final mutation in mutations) {
      if (!_validate(mutation, freshProgressOf)) {
        // Stale — discard and drop from the persisted queue.
        _queue.removeWhere((m) => m._same(mutation));
        _schedulePersist();
        continue;
      }

      if (executedThisWindow >= rateLimit.perMinute) {
        final elapsed = _clock.now().difference(windowStart);
        final wait = const Duration(minutes: 1) - elapsed;
        if (wait > Duration.zero) await _clock.sleep(wait);
        executedThisWindow = 0;
        windowStart = _clock.now();
      }

      try {
        await execute(resolveTokenFields(mutation));
        // Succeeded — leave the persisted queue.
        _queue.removeWhere((m) => m._same(mutation));
        _schedulePersist();
        executedThisWindow++;
      } on Object {
        // Keep it queued; the next flush retries.
      }
      if (rateLimit.spacing > Duration.zero) {
        await _clock.sleep(rateLimit.spacing);
      }
    }
  }

  /// Stale-write detection: an entry mutation is invalid when the server's
  /// progress already moved past the baseline recorded at queue time.
  /// delete/favourite are always valid.
  bool _validate(
    QueuedMutation mutation,
    int? Function(int mediaId)? freshProgressOf,
  ) {
    if (mutation.type != MutationType.entry) return true;
    final targetProgress = mutation.executed
        ? (mutation.result?['progress'] as num?)
        : mutation.variables['episode'] as num?;
    if (targetProgress == null) return true;
    final freshProgress = freshProgressOf?.call(mutation.mediaId);
    if (freshProgress == null) return true;
    final progressBefore = mutation.progressBefore ?? freshProgress;
    return freshProgress <= progressBefore;
  }
}
